bin/migrate-oats-data/batch_applications.py

```diff
@@ -0,0 +1,143 @@
+
+        VALUES{applications_to_insert}
+        ON CONFLICT (file_number) DO UPDATE SET
+            file_number = EXCLUDED.file_number,
+            
+            type_code = EXCLUDED.type_code,
+            applicant = EXCLUDED.applicant,
+            region_code = EXCLUDED.region_code,
+            local_government_uuid = EXCLUDED.local_government_uuid,
+            audit_created_by = EXCLUDED.audit_created_by
+    """
+
+@inject_conn_pool
+def process_applications(conn=None, batch_size=10000):
+    """
+    function uses a decorator pattern @inject_conn_pool to inject a database connection pool to the function. It fetches the total count of non duplicate applications and prints it to the console. Then, it fetches the applications to insert in batches using application IDs / file_number, constructs an insert query, and processes them.
+    """
+    with conn.cursor() as cursor:
+        cursor.execute(application_etl_temp_table())
+        cursor.execute(application_etl_insert())
+        # cursor.execute(associated_card())
+        with open(
+            "sql/insert_batch_application_count.sql", "r", encoding="utf-8"
+        ) as sql_file:
+            count_query = sql_file.read()
+            cursor.execute(count_query)
+            count_total = cursor.fetchone()[0]
+        print("- Applications to insert: ", count_total)
+
+        failed_inserts = 0
+        successful_inserts_count = 0
+        last_application_id = 0
+        
+
+        with open("sql/insert-batch-application.sql", "r", encoding="utf-8") as sql_file:
+            application_sql = sql_file.read()
+            while True:
+                cursor.execute(
+                    f"{application_sql} WHERE ae.application_id > {last_application_id} ORDER by ae.application_id"
+                )
+                
+                rows = cursor.fetchmany(batch_size)
+                if not rows:
+                    break
+                try: 
+                    applications_to_be_inserted_count = len(rows)
+
+                    insert_query = compile_application_insert_query(
+                        applications_to_be_inserted_count
+                    )
+                    cursor.execute(insert_query, rows)
+                    conn.commit()
+
+                    last_application_id = rows[-1][0]
+                    successful_inserts_count = (
+                        successful_inserts_count + applications_to_be_inserted_count
+                    )
+
+                    print(
+                        f"retrieved/inserted items count: {applications_to_be_inserted_count}; total successfully inserted/updated applications so far {successful_inserts_count}; last inserted applidation_id: {last_application_id}"
+
+                    )
+                except Exception as e:
+                    conn.rollback()
+                    print("Error", e)
+                    failed_inserts += len(rows)
```

Comment:
We discussed that the failed_inserts would show the wrong value at this point since the +=len(rows) is not equal to the number of rows after the shift in last_application_id +1. Let's replace it with the count_total-successful_inserts_count after the last batch insert.

Reply:
logic updated